src/model-definition.ts

```diff
@@ -0,0 +1,893 @@
+   */
+  readonly #dateAttributeNames = new Set<string>();
+
+  /**
+   * @deprecated Code should not rely on this as users can create custom attributes.
+   */
+  readonly dateAttributeNames = new SetView(this.#dateAttributeNames);
+
+  #autoIncrementAttributeName: string | null = null;
+  get autoIncrementAttributeName(): string | null {
+    return this.#autoIncrementAttributeName;
+  }
+
+  readonly #defaultValues = new Map</* attribute name */ string, () => unknown>();
+  readonly defaultValues = new MapView(this.#defaultValues);
+
+  /**
+   * Final list of indexes, built by {@link refreshIndexes}
+   */
+  #indexes: IndexOptions[] = [];
+
+  /**
+   * @deprecated Temporary property to be able to use elements that have not migrated to ModelDefinition yet.
+   */
+  readonly #model: ModelStatic;
+
+  get modelName(): string {
+    return this.options.modelName;
+  }
+
+  get underscored(): boolean {
+    return this.options.underscored;
+  }
+
+  get sequelize(): Sequelize {
+    return this.#sequelize;
+  }
+
+  get hooks(): HookHandler<ModelHooks> {
+    return staticModelHooks.getFor(this);
+  }
+
+  constructor(attributesOptions: ModelAttributes, modelOptions: InitOptions, model: ModelStatic) {
+    if (!modelOptions.sequelize) {
+      throw new Error('new ModelDefinition() expects a Sequelize instance to be passed through the option bag, which is the second parameter.');
+    }
+
+    if (!modelOptions.modelName) {
+      throw new Error('new ModelDefinition() expects a modelName to be passed through the option bag, which is the second parameter.');
+    }
+
+    this.#sequelize = modelOptions.sequelize;
+    this.#model = model;
+
+    const globalOptions = this.#sequelize.options;
+
+    // caution: mergeModelOptions mutates its first input
+    this.options = mergeModelOptions(
+      Object.assign(
+        // default options
+        {
+          noPrimaryKey: false,
+          timestamps: true,
+          validate: {},
+          freezeTableName: false,
+          underscored: false,
+          paranoid: false,
+          rejectOnEmpty: false,
+          schema: '',
+          schemaDelimiter: '',
+          defaultScope: {},
+          scopes: {},
+          name: {},
+          indexes: [],
+        },
+        globalOptions.define as ModelOptions,
+      ),
+      modelOptions,
+      true,
+    ) as BuiltModelOptions;
+
+    // @ts-expect-error -- guide to help users migrate to alternatives, these were deprecated in v6
+    if (this.options.getterMethods || this.options.setterMethods) {
+      throw new Error(`Error in the definition of Model ${this.modelName}: The "getterMethods" and "setterMethods" options have been removed.
+
+If you need to use getters & setters that behave like attributes, use VIRTUAL attributes.
+If you need regular getters & setters, define your model as a class and add getter & setters.
+See https://sequelize.org/docs/v6/core-concepts/getters-setters-virtuals/ for more information.`);
```

Comment:
Should we use the v7 docs for this instead? So people don't click around in the v6 docs while using v7

Reply:
I wanted to point it to https://sequelize.org/docs/v6/core-concepts/getters-setters-virtuals/#deprecated-in-sequelize-v7-gettermethods-and-settermethods, which is a paragraph that will be dropped in our v7 documentation